File: aoc14b.pl
#!/usr/bin/perl -w

use strict;

# my $goal = "793061";
my $goal = shift @ARGV;

sub report {
    print STDERR @_, "\n";
}

sub verbose {
    print STDERR @_, "\n";
}

my $init = "37";
my $scoreboard = $init;
my @e = (0, 1);

sub extend {
    my $newscore = substr($scoreboard, $e[0], 1) + substr($scoreboard, $e[1], 1);
    # verbose("newscore $newscore, length ", length($newscore));

    for (my $i = 0; $i < length($newscore); $i++) {
        $scoreboard .= substr($newscore, $i, 1);
        # verbose "scoreboard now $scoreboard"
    }

    for (my $i = 0; $i < scalar(@e); $i++) {
        # verbose "elf $i at $e[$i]";
        my $cscore = substr($scoreboard, $e[$i], 1);
        $e[$i] = ($e[$i] + $cscore + 1) % length($scoreboard);
        # verbose "now elf $i at $e[$i]";
    }
}

while (length($scoreboard) < length($goal)+1) {
    extend();
}

until ($goal eq (substr($scoreboard, length($scoreboard)-length($goal)-1, length($goal))) or
       $goal eq (substr($scoreboard, length($scoreboard)-length($goal), length($goal)))) {
    extend()
}

# this might (probably usually is) off-by-one; use the subsequent output to fix it up

report "length of scoreboard minus length of goal is ",
     length($scoreboard)-length($goal);

report "last ten on scoreboard are ", substr($scoreboard, length($scoreboard)-10);


open (my $F, ">f");
print $F $scoreboard;
close ($F)

File: aoc14b.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define NUM_ELEMENTS(a) (sizeof(a)/sizeof((a)[0]))

//#define DPRINTF(s, args...) fprintf(stderr, s "\n", args)
#define DPRINTF(s, args...) /**/

// It is a sin to use global variables.
static const char *goal;
static size_t goallen;
static const int MAX_SCOREBOARD = 1024*1024*1024;
static char *scoreboard;
static char *p;
static int e[] = { 0, 1 };         /* elf positions */

static inline int scoreboard_len() {
    return p - scoreboard;
}

void extend() {
    int newscore = (scoreboard[e[0]] - '0' +
                    scoreboard[e[1]] - '0');
    DPRINTF("newscore %d", newscore);
    int tens = newscore / 10;
    if (tens) {
        DPRINTF("tens %d", tens);
        *p = tens + '0';
        p++;
    }
    int ones = newscore % 10;
    DPRINTF("ones %d", ones);
    *p = ones + '0';
    p++;

    // extend scoreboard somewhat sloppily until it can accomidate
    // sloppy memcmp
    for (int i = 0; i < NUM_ELEMENTS(e); i++) {
        int cscore = scoreboard[e[i]] - '0';
        e[i] = (e[i] + cscore + 1) % (scoreboard_len());
    }
    DPRINTF("p-4=%s", p-4);
    DPRINTF("scoreboard is now %s", scoreboard);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "one arg plz k thx bye\n");
        return 1;
    }

    goal = argv[1];
    DPRINTF("goal %s", goal);
    goallen = strlen(goal);
    scoreboard = calloc(sizeof(char), MAX_SCOREBOARD);
    strcpy(scoreboard, "37");
    p = scoreboard + strlen(scoreboard);
    DPRINTF("p-2=%s", p-2);

    while (scoreboard_len() < goallen + 1) {
        // this might be bogus if the goal is right in the beginning.
        extend();
    }

    while (scoreboard_len() < MAX_SCOREBOARD &&
           memcmp(p - goallen, goal, goallen) &&
           memcmp(p - goallen - 1, goal, goallen)) {
        if (scoreboard_len() > 20) {
            DPRINTF("scoreboard last 20: %s\n", p - 20);
        }

        extend();

        DPRINTF("ends %d", memcmp(p - goallen, goal, goallen));
        DPRINTF("ends-1 %d", memcmp(p - goallen - 1, goal, goallen));
    }

    if (scoreboard_len() == MAX_SCOREBOARD) {
        printf("overflowed allocation\n");
    }

    if (0 == memcmp(p - goallen, goal, goallen)) {
        printf("%lu\n", scoreboard_len() - goallen);
    } else if (0 == memcmp(p - goallen - 1, goal, goallen)) {
        printf("%lu\n", scoreboard_len() - goallen - 1);
    } else {
        printf("uhoh\n");
    }

    if (scoreboard_len() < 20) {
        printf("scoreboard: %s\n", scoreboard);
    } else {
        printf("scoreboard last 20: %s\n", p - 20);
    }

    return 0;
}
